File: src/algebra1_exponent_rules_hard.py
"""
algebra1 - exponent_rules (hard)
Generated: 2026-02-22T03:58:26.371899
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 6)

    if problem_type == 1:
        # Simplify with coefficients and negative exponents
        base_var = choice([x, y])
        exp_num = randint(1, 5)
        exp_den = exp_num + randint(1, 4)
        c_num = randint(2, 9)
        c_den = randint(2, 9)

        final_coeff = Rational(c_num, c_den)
        final_exp = exp_num - exp_den  # negative

        if final_coeff == 1:
            ans = base_var**final_exp
        else:
            ans = final_coeff * base_var**final_exp

        return problem(
            question=f"Simplify: $\\frac{{{c_num}{latex(base_var)}^{{{exp_num}}}}}{{{c_den}{latex(base_var)}^{{{exp_den}}}}}$",
            answer=ans,
            difficulty=(1600, 1700),
            topic="algebra1/exponent_rules",
            solution=steps(
                f"$\\frac{{{c_num}}}{{{c_den}}} \\cdot {latex(base_var)}^{{{exp_num}-{exp_den}}} = {latex(final_coeff)} \\cdot {latex(base_var)}^{{{final_exp}}}$",
                f"$= {latex(ans)}$"
            )
        )

    elif problem_type == 2:
        # Multi-variable: (x^a * y^b)^c / (x^d * y^e)
        a_e = randint(2, 5)
        b_e = randint(2, 5)
        c_e = randint(2, 4)
        d_e = randint(1, a_e*c_e - 1)
        e_e = randint(1, b_e*c_e - 1)

        x_exp = a_e*c_e - d_e
        y_exp = b_e*c_e - e_e
        ans = x**x_exp * y**y_exp

        return problem(
            question=f"Simplify: $\\frac{{(x^{{{a_e}}}y^{{{b_e}}})^{{{c_e}}}}}{{x^{{{d_e}}}y^{{{e_e}}}}}$",
            answer=ans,
            difficulty=(1650, 1750),
            topic="algebra1/exponent_rules",
            solution=steps(
                f"Apply power rule: $(x^{{{a_e}}}y^{{{b_e}}})^{{{c_e}}} = x^{{{a_e*c_e}}}y^{{{b_e*c_e}}}$",
                f"Apply quotient rule: $x^{{{a_e*c_e}-{d_e}}}y^{{{b_e*c_e}-{e_e}}} = {latex(ans)}$"
            )
        )

    elif problem_type == 3:
        # Exponential equation: base^(linear1) = base^(linear2)
        base_val = choice([2, 3, 4, 5, 7])
        # Random coefficients for both sides
        a_coeff = randint(2, 6)
        b_coeff = randint(1, 5)
        c_coeff = randint(3, 8)
        while c_coeff == a_coeff:
            c_coeff = randint(3, 8)
        # a_coeff*x + b_const = c_coeff*x + d_const → pick x_val and derive
        x_val = randint(1, 10)
        lhs_exp = a_coeff * x_val + b_coeff
        d_const = lhs_exp - c_coeff * x_val

        ans = x_val
        return problem(
            question=f"Solve for $x$: ${base_val}^{{{a_coeff}x+{b_coeff}}} = {base_val}^{{{c_coeff}x+{d_const}}}$",
            answer=ans,
            difficulty=(1600, 1700),
            topic="algebra1/exponent_rules",
            solution=steps(
                f"Since bases are equal, set exponents equal: ${a_coeff}x + {b_coeff} = {c_coeff}x + {d_const}$",
                f"${b_coeff} - {d_const} = {c_coeff}x - {a_coeff}x$",
                f"${b_coeff - d_const} = {c_coeff - a_coeff}x$",
                f"$x = {ans}$"
            )
        )

    elif problem_type == 4:
        # (x^a)^c * x^(-b) - power then product with negative exponent
        a_e = randint(2, 5)
        c_e = randint(2, 4)
        b_e = randint(1, a_e*c_e - 1)
        final_exp = a_e*c_e - b_e
        ans = x**final_exp

        return problem(
            question=f"Simplify: $(x^{{{a_e}}})^{{{c_e}}} \\cdot x^{{-{b_e}}}$",
            answer=ans,
            difficulty=(1600, 1700),
            topic="algebra1/exponent_rules",
            solution=steps(
                f"Power rule: $(x^{{{a_e}}})^{{{c_e}}} = x^{{{a_e*c_e}}}$",
                f"Product rule: $x^{{{a_e*c_e}}} \\cdot x^{{-{b_e}}} = x^{{{a_e*c_e}+(-{b_e})}} = x^{{{final_exp}}}$"
            )
        )

    elif problem_type == 5:
        # Complex: (x^a / x^b)^c * (x^d)^e
        a_e = randint(3, 6)
        b_e = randint(a_e+1, a_e+4)
        c_e = randint(2, 4)
        d_e = randint(1, 4)
        e_e = randint(2, 4)

        final_exp = (a_e - b_e)*c_e + d_e*e_e
        ans = x**final_exp

        return problem(
            question=f"Simplify: $\\left(\\frac{{x^{{{a_e}}}}}{{x^{{{b_e}}}}}\\right)^{{{c_e}}} \\cdot (x^{{{d_e}}})^{{{e_e}}}$",
            answer=ans,
            difficulty=(1700, 1850),
            topic="algebra1/exponent_rules",
            solution=steps(
                f"Quotient inside: $\\frac{{x^{{{a_e}}}}}{{x^{{{b_e}}}}} = x^{{{a_e-b_e}}}$",
                f"Power rule: $(x^{{{a_e-b_e}}})^{{{c_e}}} = x^{{{(a_e-b_e)*c_e}}}$",
                f"Other term: $(x^{{{d_e}}})^{{{e_e}}} = x^{{{d_e*e_e}}}$",
                f"Product: $x^{{{(a_e-b_e)*c_e}}} \\cdot x^{{{d_e*e_e}}} = x^{{{final_exp}}}$"
            )
        )

    else:
        # Fractional exponent: x^(1/n) = n-th root
        base_var = choice([x, y])
        n = randint(2, 5)
        m = randint(1, n)
        # Simplify x^(m/n)
        ans = base_var**Rational(m, n)
        # Sometimes ask to evaluate: base^(m/n) for numeric base
        base_num = choice([4, 8, 9, 16, 25, 27, 32, 64])
        from sympy import Rational as R
        exp_frac = R(m, n)
        val = base_num**exp_frac
        if val.is_Integer or val.is_Rational:
            return problem(
                question=f"Evaluate: ${base_num}^{{\\frac{{{m}}}{{{n}}}}}$",
                answer=val,
                difficulty=(1650, 1750),
                topic="algebra1/exponent_rules",
                solution=steps(
                    f"$x^{{m/n}} = (\\sqrt[n]{{x}})^m = (x^{{1/n}})^m$",
                    f"$= (\\sqrt[{n}]{{{base_num}}})^{{{m}}}$",
                    f"$= {latex(val)}$"
                )
            )
        else:
            # Fall back to variable simplification
            exp1 = randint(2, 5)
            exp2 = randint(2, 5)
            ans2 = base_var**(exp1+exp2)
            return problem(
                question=f"Simplify: ${latex(base_var)}^{{{exp1}}} \\cdot {latex(base_var)}^{{{exp2}}}$",
                answer=ans2,
                difficulty=(1600, 1700),
                topic="algebra1/exponent_rules",
                solution=steps(
                    f"$= {latex(base_var)}^{{{exp1+exp2}}}$"
                )
            )

emit(generate())

File: src/algebra1_exponent_rules_hard.jl
# algebra1 - exponent_rules (hard)
# Generated: 2026-03-08T19:58:48.473851

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y begin
    set_topic!("algebra1/exponent_rules")
    
    problem_type = choice(1:6)
    
    if problem_type == 1
        # Simplify product of powers with same base
        base = choice([2, 3, 5, 7])
        exp1 = randint(3, 8)
        exp2 = randint(3, 8)
        exp3 = randint(2, 6)
        total_exp = exp1 + exp2 + exp3
        
        problem(
            question="Simplify \\($(base)^{$(exp1)} \\cdot $(base)^{$(exp2)} \\cdot $(base)^{$(exp3)}\\) and express as a single power.",
            answer="$(base)^$(total_exp)",
            difficulty=(1800, 2000),
            solution=steps(
                "Given: \\($(base)^{$(exp1)} \\cdot $(base)^{$(exp2)} \\cdot $(base)^{$(exp3)}\\)",
                "Apply the product rule: \\(a^m \\cdot a^n = a^{m+n}\\)",
                "Add exponents: \\($(exp1) + $(exp2) + $(exp3) = $(total_exp)\\)",
                sol("Answer", "$(base)^{$(total_exp)}")
            ),
            time=90
        )
        
    elseif problem_type == 2
        # Power of a power with algebraic expressions
        coef = nonzero(-5, 5)
        inner_exp = randint(2, 5)
        outer_exp = randint(2, 4)
        final_exp = inner_exp * outer_exp
        final_coef = coef^outer_exp
        
        problem(
            question="Simplify \\(($(coef)x^{$(inner_exp)})^{$(outer_exp)}\\) and express in the form \\(ax^n\\).",
            answer="$(final_coef)x^$(final_exp)",
            difficulty=(1900, 2100),
            solution=steps(
                "Given: \\(($(coef)x^{$(inner_exp)})^{$(outer_exp)}\\)",
                "Apply the power rule: \\((ab)^n = a^n b^n\\) and \\((a^m)^n = a^{mn}\\)",
                "Coefficient: \\($(coef)^{$(outer_exp)} = $(final_coef)\\)",
                "Exponent: \\($(inner_exp) \\cdot $(outer_exp) = $(final_exp)\\)",
                sol("Answer", "$(final_coef)x^{$(final_exp)}")
            ),
            time=120
        )
        
    elseif problem_type == 3
        # Quotient of powers with simplification
        base = choice([2, 3, 5])
        numerator_exp = randint(8, 15)
        denominator_exp = randint(4, 8)
        result_exp = numerator_exp - denominator_exp
        
        problem(
            question="Simplify \\(\\frac{$(base)^{$(numerator_exp)}}{$(base)^{$(denominator_exp)}}\\) and express as a single power.",
            answer="$(base)^$(result_exp)",
            difficulty=(1850, 2050),
            solution=steps(
                "Given: \\(\\frac{$(base)^{$(numerator_exp)}}{$(base)^{$(denominator_exp)}}\\)",
                "Apply the quotient rule: \\(\\frac{a^m}{a^n} = a^{m-n}\\)",
                "Subtract exponents: \\($(numerator_exp) - $(denominator_exp) = $(result_exp)\\)",
                sol("Answer", "$(base)^{$(result_exp)}")
            ),
            time=90
        )
        
    elseif problem_type == 4
        # Complex expression with multiple variables
        a_exp = randint(3, 7)
        b_exp = randint(2, 6)
        c_exp = randint(2, 5)
        power = randint(2, 4)
        
        final_a = a_exp * power
        final_b = b_exp * power
        final_c = c_exp * power
        
        problem(
            question="Simplify \\((x^{$(a_exp)} y^{$(b_exp)} z^{$(c_exp)})^{$(power)}\\) and express in the form \\(x^a y^b z^c\\).",
            answer="x^$(final_a)*y^$(final_b)*z^$(final_c)",
            difficulty=(2000, 2200),
            solution=steps(
                "Given: \\((x^{$(a_exp)} y^{$(b_exp)} z^{$(c_exp)})^{$(power)}\\)",
                "Apply the power rule to each variable: \\((a^m)^n = a^{mn}\\)",
                "\\(x: $(a_exp) \\cdot $(power) = $(final_a)\\)",
                "\\(y: $(b_exp) \\cdot $(power) = $(final_b)\\)",
                "\\(z: $(c_exp) \\cdot $(power) = $(final_c)\\)",
                sol("Answer", "x^{$(final_a)} y^{$(final_b)} z^{$(final_c)}")
            ),
            time=150
        )
        
    elseif problem_type == 5
        # Negative and zero exponents with fractions
        base_num = choice([2, 3, 5, 7])
        base_den = choice([2, 3, 5, 7])
        while base_den == base_num
            base_den = choice([2, 3, 5, 7])
        end
        
        exp = randint(3, 6)
        ans_num = base_den^exp
        ans_den = base_num^exp
        
        problem(
            question="Simplify \\(\\left(\\frac{$(base_num)}{$(base_den)}\\right)^{-$(exp)}\\) and express as a simplified fraction.",
            answer="$(ans_num)/$(ans_den)",
            difficulty=(2100, 2300),
            solution=steps(
                "Given: \\(\\left(\\frac{$(base_num)}{$(base_den)}\\right)^{-$(exp)}\\)",
                "Apply the negative exponent rule: \\(a^{-n} = \\frac{1}{a^n}\\)",
                "This gives: \\(\\left(\\frac{$(base_den)}{$(base_num)}\\right)^{$(exp)}\\)",
                "Apply power to numerator and denominator: \\(\\frac{$(base_den)^{$(exp)}}{$(base_num)^{$(exp)}} = \\frac{$(ans_num)}{$(ans_den)}\\)",
                sol("Answer", "\\frac{$(ans_num)}{$(ans_den)}")
            ),
            time=180
        )
        
    else  # problem_type == 6
        # Multi-step simplification combining rules
        coef1 = nonzero(-4, 4)
        coef2 = nonzero(-4, 4)
        exp1 = randint(4, 8)
        exp2 = randint(3, 7)
        exp3 = randint(2, 4)
        
        final_coef = coef1 * coef2^exp3
        final_exp = exp1 + exp2 * exp3
        
        problem(
            question="Simplify \\($(coef1)x^{$(exp1)} \\cdot ($(coef2)x^{$(exp2)})^{$(exp3)}\\) and express in the form \\(ax^n\\).",
            answer="$(final_coef)x^$(final_exp)",
            difficulty=(2200, 2400),
            solution=steps(
                "Given: \\($(coef1)x^{$(exp1)} \\cdot ($(coef2)x^{$(exp2)})^{$(exp3)}\\)",
                "First simplify the power: \\(($(coef2)x^{$(exp2)})^{$(exp3)} = $(coef2^exp3)x^{$(exp2 * exp3)}\\)",
                "Now multiply: \\($(coef1)x^{$(exp1)} \\cdot $(coef2^exp3)x^{$(exp2 * exp3)}\\)",
                "Combine coefficients: \\($(coef1) \\cdot $(coef2^exp3) = $(final_coef)\\)",
                "Add exponents: \\($(exp1) + $(exp2 * exp3) = $(final_exp)\\)",
                sol("Answer", "$(final_coef)x^{$(final_exp)}")
            ),
            time=200
        )
    end
end